src/core/storage/fileio/s3_filesys.hpp

```diff
@@ -0,0 +1,384 @@
+ public:
+  virtual ~AWSReadStreamBase() {
+    logstream(LOG_DEBUG) << "~AWSReadStream" << std::endl;
+    Close();
+  }
+
+  virtual void Close() {
+    logstream(LOG_DEBUG) << "AWSReadStream::Close()" << std::endl;
+    Reset(file_size_);
+  }
+
+  virtual size_t Tell(void) { return curr_bytes_; }
+
+  virtual size_t FileSize() const { return file_size_; }
+
+  virtual bool AtEnd(void) const { return curr_bytes_ == file_size_; }
+
+  virtual void Write(const void *ptr, size_t size) {
+    std_log_and_throw(std::runtime_error,
+                      "AWSReadStreamBase is not supposed to write");
+  }
+  // lazy seek function
+  virtual void Seek(size_t pos) {
+    ASSERT_TRUE(pos < file_size_);
+    if (curr_bytes_ != pos) {
+      this->Reset(pos);
+    }
+  }
+  virtual size_t Read(void *ptr, size_t size);
+
+ protected:
+  AWSReadStreamBase() : file_size_(0), read_ptr_(0), curr_bytes_(0) {}
+
+  /*!
+   * \brief initialize the ecurl request,
+   * \param begin_bytes the beginning bytes of the stream
+   */
+  virtual void InitRequest(size_t begin_bytes, s3url &url) = 0;
+
+ protected:
+  // the total size of the file
+  size_t file_size_ = 0;
+
+  void SetBegin(size_t begin_bytes) {
+    ASSERT_TRUE(begin_bytes < file_size_);
+    curr_bytes_ = begin_bytes;
+  }
+
+ protected:
+  s3url url_;
+
+ private:
+  /*!
+   * \brief called by child class to initialize read
+   * \param begin_bytes the beginning bytes of the stream
+   */
+  void Reset(size_t begin_bytes);
+
+  /*!
+   * \brief try to fill the buffer with at least wanted bytes
+   * \param want_bytes number of bytes we want to fill
+   * \return number of remainning running curl handles
+   */
+
+  int FillBuffer(size_t want_bytes);
+  // data buffer
+  std::string buffer_;
+  // data pointer to read position
+  size_t read_ptr_;
+  // current position in the stream
+  size_t curr_bytes_;
+};
+
+/*! \brief reader stream that can be used to read */
+class ReadStream : public AWSReadStreamBase {
+ public:
+  ReadStream(const s3url &url, size_t file_size) {
+    file_size_ = file_size;
+    url_ = url;
+  }
+
+  virtual ~ReadStream() {}
+
+ protected:
+  // implement InitRequest
+  virtual void InitRequest(size_t begin_bytes, s3url &url) {
+    SetBegin(begin_bytes);
+    url_ = url;
+  }
+};
+
+constexpr size_t S3_MIN_MULTIPART_SIZE = 5 * 1024 * 1024;  // 5MB
+
+class WriteStream : public Stream {
+ public:
+  WriteStream(const s3url &url, bool no_exception = false)
+      : url_(url), no_exception_(no_exception) {
+    const char *buz = getenv("TURI_S3_WRITE_BUFFER_MB");
+    if (buz != nullptr) {
+      max_buffer_size_ = std::max(static_cast<size_t>(atol(buz)) << 20UL,
+                                  S3_MIN_MULTIPART_SIZE);
+    } else {
+      // 64 MB
+      const size_t kDefaultBufferSize = 64 << 20UL;
+      max_buffer_size_ = kDefaultBufferSize;
+    }
+
+    InitRequest(url_);
+  }
+
+  virtual size_t Read(void *ptr, size_t size) {
+    if (no_exception_) {
```

Comment:
I don't understand this no_exception logic.

Reply:
this is the logic inherits from dlmcio code. Some data stream impl requires no exception. if the size is returned as 0, it means no data from stream. And the user should check the io bits such as eof, bad, to determine their error handling logic.

Reply:
I don't understand why the logic is if(no_exception) { // throw exception }

Reply:
I see! Let me check. might be my sloppiness.

Reply:
yeah, forget one exclamation sign! Nice catch!